### PR TITLE
Properly receive the MCP POST URL from the MCP server

### LINKED ISSUE #2266: [BUG] MCP: obtain the POST url from the server
Currently, we have the POST url configurable, but it's actually something that the server gives you during the initialization, and it includes the proper session ID.

### PR DESCRIPTION
**Pull request by jmartisk**
Closes #2266

### CODE REVIEW

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java [hunk 1]
```diff
@@ -116,18 +122,32 @@ private JsonNode executeAndWait(Request request, Long id) {
...
-        CompletableFuture<Void> initializationFinished = new CompletableFuture<>();
+        CompletableFuture<String> initializationFinished = new CompletableFuture<>();
         SseEventListener listener = new SseEventListener(pendingOperations, logResponses, initializationFinished);
         EventSource eventSource = EventSources.createFactory(client).newEventSource(request, listener);
-        // wait for the SSE channel to be created, throw an exception if that failed
+        // wait for the SSE channel to be created, receive the POST url from the server, throw an exception if that
+        // failed
         try {
-            initializationFinished.get();
+            int timeout = client.callTimeoutMillis() > 0 ? client.callTimeoutMillis() : Integer.MAX_VALUE;
```
Comment (langchain4j) [inline langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: Is it a good idea to have `MAX_VALUE` as a timeout?

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java [hunk 2]
```diff
@@ -116,18 +122,32 @@ private JsonNode executeAndWait(Request request, Long id) {
...
+            log.debug("Received the server's POST URL: " + postUrl);
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
         return eventSource;
     }
 
+    private String buildAbsolutePostUrl(String relativePostUrl) {
+        try {
+            URL sseUrl = new URL(this.sseUrl);
```
Comment (langchain4j) [inline langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java]: Perharps we should use `URI` here instead?

**Review by ghost — COMMENTED** [review]
@jmartisk thanks a lot!

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 1 above)
Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: I could instead create an `if` statement and call the `get()` method without a timeout, the code will look a bit more complicated, but the behavior will basically be the same.

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 2 above)
Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java]: Like this?
```
URI sseUrl = new URI(this.sseUrl);
URL postUrl = new URL(sseUrl.toURL(), relativePostUrl);
return postUrl.toString();
```

or what else do you have in mind?

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 1 above)
Comment (langchain4j) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: I meant that maybe we should specify a bit more meaningful default value for the timeout, perhaps 60-90 seconds? `MAX_VALUE` seems like "wait for it forever"

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java]: I have heard bad things about `URL` (how it initiates a connection in `equals()`, etc.).
IIUC, the general advice is to use `URI` instead.
So if this is true, I assume we should use URI everywhere?

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 1 above)
Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: The default is 60 seconds (this is just to handle the case that somebody explicitly specifies the timeout to be 0)

Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: Ah, hold on, you may actually be right that we don't have a default. Ok I'll add a default as well, but we have to keep handling the case that somebody specifies it to be zero explicitly (which should mean "no timeout", but calling `completableFuture.get(0)` would throw an exception immediately)

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 2 above)
Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java]: Ok, I've replaced it with `URI.create(this.sseUrl).resolve(relativePostUrl).toString()`

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 1 above)
Comment (jmartisk) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: I've added a 60 second timeout, right now it's just one value for callTimeout, connectTimeout, readTimeout, writeTimeout. Do you prefer having four separate values?

Comment (langchain4j) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java:131]: @jmartisk thank you! I think it is ok to have a single value for now

File: langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java]: Thanks!

**Review by ghost — APPROVED** [review]
@jmartisk thank you!

**Pull request merged**